src/libcollections/slice.rs

```diff
@@ -1305,213 +1304,290 @@ impl<T: Clone> ToOwned for [T] {
-    // but hasn't been tested/tuned extensively
-    let insertion = if size_of::<T>() <= 16 {
-        BASE_INSERTION
+    let len = v.len();
+    let v = v.as_mut_ptr();
+    let v_mid = v.offset(mid as isize);
+    let v_end = v.offset(len as isize);
+
+    // The merge process first copies the shorter run into `buf`. Then it traces the newly copied
+    // run and the longer run forwards (or backwards), comparing their next unconsumed elements and
+    // copying the lesser (or greater) one into `v`.
+    //
+    // As soon as the shorter run is fully consumed, the process is done. If the longer run gets
+    // consumed first, then we must copy the remains of the shorter run into `v` at the place where
+    // we left off. The guard will do that as soon as it's dropped.
+    //
+    // Panic safety:
+    //
+    // The guard holds the intermediate state of the merge process. It knows about the still
+    // unconsumed range in `buf` and the position in `v` where the next object is about to be
+    // placed. If `compare` panics at any point during the process, the guard will get dropped,
+    // copy the range back into `v`, and thus ensure that `v` still holds every object it
+    // initially held exactly once.
+    let mut guard;
+
+    if mid <= len - mid {
+        // The left run is shorter.
+        ptr::copy_nonoverlapping(v, buf, mid);
+        guard = Guard {
+            start: buf,
+            end: buf.offset(mid as isize),
+            dst: v,
+        };
+
+        let left = &mut guard.start;
+        let mut right = v_mid;
+        let out = &mut guard.dst;
+
+        while *left < guard.end && right < v_end {
+            // Consume the lesser side.
+            // If equal, prefer the left run to maintain stability.
+            let to_copy = if compare(&**left, &*right) == Greater {
+                forward(&mut right)
+            } else {
+                forward(left)
+            };
+            ptr::copy_nonoverlapping(to_copy, forward(out), 1);
+        }
     } else {
-        LARGE_INSERTION
-    };
+        // The right run is shorter.
+        ptr::copy_nonoverlapping(v_mid, buf, len - mid);
+        guard = Guard {
+            start: buf,
+            end: buf.offset((len - mid) as isize),
+            dst: v_mid,
+        };
+
+        let left = &mut guard.dst;
+        let right = &mut guard.end;
+        let mut out = v_end;
+
+        while v < *left && buf < *right {
+            // Consume the greater side.
+            // If equal, prefer the right run to maintain stability.
+            let to_copy = if compare(&*left.offset(-1), &*right.offset(-1)) == Greater {
+                backward(left)
+            } else {
+                backward(right)
+            };
+            ptr::copy_nonoverlapping(to_copy, backward(&mut out), 1);
+        }
+    }
 
-    let len = v.len();
+    // Increments the pointer and returns the old value.
+    unsafe fn forward<T>(ptr: &mut *mut T) -> *mut T {
```

Comment:
`forward` is also known as `ptr++` in C, that is "post increment".

In my opinion, this is in easier to read style with an extension method on raw pointers, so that it reads `ptr.post_increment()`. Not sure what others think?